include/circt/Dialect/Calyx/CalyxLoweringUtils.h

```diff
@@ -89,41 +131,303 @@ class WhileOpInterface {
+// several lowering patterns.
+template <typename Loop>
+class LoopLoweringStateInterface {
+  static_assert(std::is_base_of_v<LoopInterface, Loop>);
 
-// Returns true if there exists only a single memref::LoadOp which loads from
-// the memory referenced by loadOp.
-bool singleLoadFromMemory(Value memoryReference);
+public:
+  ~LoopLoweringStateInterface() = default;
 
-// Returns true if there are no memref::StoreOp uses with the referenced
-// memory.
-bool noStoresToMemory(Value memoryReference);
+  /// Register reg as being the idx'th iter_args register for 'op'.
+  void addLoopIterReg(Loop op, calyx::RegisterOp reg, unsigned idx) {
+    assert(loopIterRegs[op.getOperation()].count(idx) == 0 &&
+           "A register was already registered for the given loop iter_arg "
+           "index");
+    assert(idx < op.getBodyArgs().size());
+    loopIterRegs[op.getOperation()][idx] = reg;
+  }
 
-// Get the index'th output port of compOp.
-Value getComponentOutput(calyx::ComponentOp compOp, unsigned outPortIdx);
+  /// Return a mapping of block argument indices to block argument.
+  calyx::RegisterOp getLoopIterReg(Loop op, unsigned idx) {
+    auto iterRegs = getLoopIterRegs(op);
+    auto it = iterRegs.find(idx);
+    assert(it != iterRegs.end() &&
+           "No iter arg register set for the provided index");
+    return it->second;
+  }
 
-// If the provided type is an index type, converts it to i32, else, returns the
-// unmodified type.
-Type convIndexType(PatternRewriter &rewriter, Type type);
+  /// Return a mapping of block argument indices to block argument.
+  const DenseMap<unsigned, calyx::RegisterOp> &getLoopIterRegs(Loop op) {
+    return loopIterRegs[op.getOperation()];
+  }
 
-// Creates a new calyx::CombGroupOp or calyx::GroupOp group within compOp.
-template <typename TGroup>
-TGroup createGroup(PatternRewriter &rewriter, calyx::ComponentOp compOp,
-                   Location loc, Twine uniqueName) {
-  mlir::IRRewriter::InsertionGuard guard(rewriter);
-  rewriter.setInsertionPointToEnd(compOp.getWiresOp().getBody());
-  return rewriter.create<TGroup>(loc, uniqueName.str());
-}
+  /// Registers grp to be the loop latch group of `op`.
+  void setLoopLatchGroup(Loop op, calyx::GroupOp group) {
+    Operation *operation = op.getOperation();
+    assert(loopLatchGroups.count(operation) == 0 &&
+           "A latch group was already set for this loopOp");
+    loopLatchGroups[operation] = group;
+  }
+
+  /// Retrieve the loop latch group registered for `op`.
+  calyx::GroupOp getLoopLatchGroup(Loop op) {
+    auto it = loopLatchGroups.find(op.getOperation());
+    assert(it != loopLatchGroups.end() &&
+           "No loop latch group was set for this loopOp");
+    return it->second;
+  }
+
+  /// Creates register assignment operations within the provided groupOp.
+  /// The component operation will house the constants.
+  void buildAssignmentsForRegisterWrite(PatternRewriter &rewriter,
```

Comment:
Can be moved to a free function (using `OpBuilder` instead of a `PatternRewriter`).

Reply:
I've moved it as a free function. Doesn't it need PatternRewriter given its use case?

Reply:
As far as i can tell, none of the PatternRewriter API is being used, so I don't see any reason to restrict it. PatternRewriter is a specialization of OpBuilder that is only created for and can be used by rewrite patterns.